Get Catalog API running

jaredkosanovic opened the pull request:
This PR is meant to merge the class-search, course-subjects, and terms APIs into one application and update the code with the new APIs from the backend ERP.

magical committed:
Explicitly register LoggingExceptionMapper

CO-709

Dropwizard's LoggingExceptionMapper, which normally handles unexpected
exceptions, was getting overridden by our GenericExceptionMapper.
That is intended. But it turns out LoggingExceptionMapper is also
responsible for turning WebApplicationExceptions into nice error
messages with the proper HTTP status codes. So subclasses of
WebApplicationException like NotAllowedException, which should generate
a 405 Method Not Allowed response, were instead getting caught by
GenericExceptionMapper and turned into a 500 Internal Server Error.

To work around this, explicitly register
LoggingExceptionMapper<WebApplicationException> to the jersey
application.

tsoliangwu0130 committed:
Add 409 response

tsoliangwu0130 committed:
Add PATCH class

tsoliangwu0130 committed:
Merge pull request #55 from osu-mist/feature/CO-701-add-PATCH-annotation-class

CO-701 Add PATCH class

tsoliangwu0130 committed:
Merge pull request #54 from osu-mist/feature/CO-700-add-409-response

CO-700 Add 409 response

magical committed:
Merge pull request #53 from osu-mist/feature/CO-709-loggingexceptionmapper

Explicitly register LoggingExceptionMapper

magical committed:
Create Application class

CO-748

We've started to accumulate a lot of common startup logic in the
SkeletonApplication class. But since each application has its own
FooApplication class, and deletes the SkeletonApplication class,
changes in SkeletonApplication are not picked up when the skeleton
is merged back into each app. The changes have to be made manually,
which in practice means they are not made at all.

Introduce a new edu.oregonstate.mist.api.Application class to act
as a base class for all our applications. The setup method performs
all the common application startup logic. Each application should call
setup at the start of their run method

Commit:
pass endpoint to url builder

Commit:
falsey falsy false falsely

Commit:
ignore if endpointUri ends with /

jaredkosanovic committed:
Merge pull request #56 from osu-mist/bug/CO-747-getpaginationurl-request-uri

pass endpoint to url builder

magical committed:
Merge pull request #57 from osu-mist/feature/CO-748-application-class

Create Application class

magical committed:
Add Error constructors

CO-706

Add static methods to the Error class to construct common Error values.
Rewrite the Resource methods in terms of the new Error methods.

Error is now responsible for loading the resource properties, which
means we can get rid of Resource.loadProperties(). Error uses a static
initializer block to load the resource properties, so there is no need
for the application to do anything to load the error text.

I considered leaving Resource.properties and Resource.loadProperties()
in place for backwards compatibility, but I'm pretty sure none of our
APIs use Resource.properties directly.

magical committed:
Merge pull request #58 from osu-mist/feature/CO-706-errors

Add Error constructors

jaredkosanovic committed:
update to 1.1.0

jaredkosanovic committed:
Merge pull request #59 from osu-mist/feature/CO-753-update-dw1.1.0

update to 1.1.0

jaredkosanovic committed:
add functions, variables, for MAX_PAGE_SIZE

jaredkosanovic committed:
Merge pull request #60 from osu-mist/feature/co-801-locations-services-bugs

add functions, variables, for MAX_PAGE_SIZE

jaredkosanovic committed:
don't make it final so we can overwrite

jaredkosanovic committed:
Merge pull request #61 from osu-mist/feature/co-801-locations-services-bugs

don't make it final so we can overwrite

jaredkosanovic committed:
use the name of the file that jenkins/api VMs expect

jaredkosanovic committed:
Merge pull request #62 from osu-mist/feature/CO-867

use the name of the file that jenkins/api VMs expect

magical committed:
ignore page numbers and sizes equal to zero

CO-949

If a user gives a page number or size equal to zero, ignore it and use
the default page number or size.

Page size must not be zero because that can lead to division by zero
errors when we calculate pagination. (CO-949)

Page number must not be zero because we consider it to be 1-indexed, and
going below that doesn't make sense.

magical committed:
update unit test comments

CO-949

These two test cases (negative page number and negative page size) were
fixed ages ago in CO-619. Update the comments to match.

magical committed:
Merge pull request #63 from osu-mist/feature/CO-949-page-size-zero

CO-949: ignore page size of zero

jaredkosanovic committed:
update skeleton to dw 1.2.4

jaredkosanovic committed:
Merge pull request #64 from osu-mist/feature/CO-990-update-skeleton

update skeleton to dw 1.2.4

jaredkosanovic committed:
add accepted response builder

jaredkosanovic committed:
Merge pull request #65 from osu-mist/feature/CO-1082-add-accepted-response

add accepted response builder

tsoliangwu0130 committed:
Update reademe

tsoliangwu0130 committed:
Update reademe

tsoliangwu0130 committed:
Update reademe

tsoliangwu0130 committed:
Merge pull request #66 from osu-mist/feature/CO-460-readme-update

Update reademe

jaredkosanovic committed:
update ojdbc driver

jaredkosanovic committed:
Merge pull request #67 from osu-mist/bug/CO-1145-encrypt-oracle-db-connections

update ojdbc driver

jaredkosanovic committed:
add forbidden response

this code originally came from the osu-events-api

jaredkosanovic committed:
add noContent (204) response

Commit:
build and runtime information should be in a meta member

wilsonia committed:
create InfoResultObject for wrapping meta information

wilsonia committed:
create InfoResultObject for wrapping meta information

wilsonia committed:
resolve merge conflicts

wilsonia committed:
create ErrorResultObject to wrap errors in a top-level array

wilsonia committed:
error fields match JSON spec

wilsonia committed:
bring all error objects up to JSON spec

wilsonia committed:
bring link object up to JSON spec

wilsonia committed:
update documentation

wilsonia committed:
default detail values

wilsonia committed:
update version in example curl request

jaredkosanovic committed:
Merge pull request #68 from osu-mist/feature/CO-1163-test-message-api

Add forbidden (403) and no content (204) responses

wilsonia committed:
remove unnecessary default empty message values

wilsonia committed:
add forbidden error

wilsonia committed:
Merge branch 'master' into feature/CO-1136-follow-JSON-spec

wilsonia committed:
update version

wilsonia committed:
Merge pull request #70 from osu-mist/feature/CO-1136-follow-JSON-spec

Feature/co 1136 follow json spec

wilsonia committed:
forbidden error in error collection

wilsonia committed:
default value for parameter so calls made without don't fail

wilsonia committed:
Merge pull request #72 from osu-mist/bug/CO-1136-builders-need-default-argument

Bug/co 1136 builders need default argument

tsoliangwu0130 committed:
Add auth header

tsoliangwu0130 committed:
Modify description

tsoliangwu0130 committed:
Merge pull request #73 from osu-mist/feature/CO-1229-add-auth-headeer

CO-1229 Add auth header

jaredkosanovic committed:
spec draft

tsoliangwu0130 commented on swagger.yaml:
```diff
@@ -35,25 +42,108 @@ paths:
+          description: "Successful response"
+          schema:
+            $ref: "#/definitions/CoursesResultObject"
+        "404":
```
will this ever return a `404`? Same with `GET /subject`, `GET /terms`

jaredkosanovic replied:
good point, they won't return 404's

tsoliangwu0130 commented on swagger.yaml:
```diff
@@ -62,209 +152,368 @@ paths:
+                meetsSaturday:
+                  description: Whether this meeting time meets on Saturday.
+                  type: boolean
+                  example: false
```
No need to change, but what do you think if we return a `meetsDays` which is an array containing the days to meet? For example: `[1, 3, 5]` stands for meeting on Monday, Wednesday and Friday?

jaredkosanovic replied:
Good suggestion. I thought about this for a bit actually. I tried to find a standard for representing a weekly schedule in a JSON API. All I could find is jCal, which doesn't really seem like it would fit for this use case. https://tools.ietf.org/html/rfc7265

I think I'll keep the various meetsSunday, meetsMonday, etc. fields as is. The array idea is good and definitely lighter weight, but I prefer doing it more explicitly here. For example, if someone looked at the API response without knowing the documentation, how would they know that 1=Monday, 2=Tuesday, etc.? Do weeks start at Sunday or Monday? Do weeks start at 0 or 1? As far as I can tell (and please let me know if I'm overlooking something), there isn't really a standard for representing the days of the week as integers.

tsoliangwu0130 replied:
I was thinking to follow the way which the cron schedule uses to represent weekdays (0-6, 0 stands for Sunday, 1 for Monday, etc.), but yeah it doesn't mean is a standard at all. Actually, it is not necessary to use integers to represent weekdays, we can even use a string array `['Monday', 'Wednesday', 'Friday']` for it. In that way, it should be more explicitly and also lighter.

jaredkosanovic replied:
@tsoliangwu0130 ahh, good idea. Going one step further, we could use the same abbreviations that the catalog uses: M, T, W, Th, F, Sa, Su

tsoliangwu0130 replied:
sounds good 👍

jaredkosanovic committed:
remove 404 responses, change /courses 404 to 400

jaredkosanovic committed:
remove fields specific to a student

the grading mode and registration status fields were specific to a student. The data in this api should be agnostic to a specific registration. I also added scheduleType and waitlist fields

jaredkosanovic committed:
display weekly schedule as array

Instead of having a separate field for each day of the week a course meets, display it as a single field with an array of day abbreviations

jaredkosanovic committed:
Merge remote-tracking branch 'skeleton/master' into feature/CO-1180-get-catalog-api-running

jaredkosanovic committed:
change to classes API

this api is meant to look up class schedule information, so it doesn't really fit the definition of a course catalog.

jaredkosanovic committed:
wip: initial refactor of application code

jaredkosanovic committed:
WIP finish application class refactor

jaredkosanovic committed:
wip: implementing /terms endpoint

jaredkosanovic committed:
finish terms endpoint

jaredkosanovic committed:
implement subjects endpoints

jaredkosanovic committed:
remove old core classes

jaredkosanovic committed:
wip: add class search endpoint

jaredkosanovic committed:
partly implement class-search endpoint

used these methods and objects from the students API which has a similar endpoint. Waiting on ellucian to fix the registration class search endpoint before proceeding, because the structure of their response might change after they update their API

jaredkosanovic committed:
remove utilhttp

jaredkosanovic committed:
add unit tests

jaredkosanovic committed:
update config example
